src/metabase/models/permissions.clj

```diff
@@ -493,36 +493,49 @@
   (classloader/require 'metabase.models.collection)
   ((resolve 'metabase.models.collection/is-personal-collection-or-descendant-of-one?) collection))
 
-(mu/defn ^:private check-not-personal-collection-or-descendant
-  "Check whether `collection-or-id` refers to a Personal Collection; if so, throw an Exception. This is done because we
-  *should* never be editing granting/etc. permissions for *Personal* Collections to entire Groups! Their owner will
-  get implicit permissions automatically, and of course admins will be able to see them,but a whole group should never
-  be given some sort of access."
+(defn- is-trash-or-descendant? [collection]
+  (classloader/require 'metabase.models.collection)
+  ((resolve 'metabase.models.collection/is-trash-or-descendant?) collection))
```

Comment:
Yeah, agreed on all counts. I would have refactored to avoid the circular dependency if I was the only thing using the pattern, but yeah, since it's already there and we'll probably be revisiting this code pretty soon, I figured it wasn't making things worse.